pmda/leaflet.py

```diff
@@ -99,62 +99,66 @@ def _find_connected_components(self, data, cutoff=15.0):
 
         """
         # pylint: disable=unsubscriptable-object
-        window, index = data[0]
-        num = window[0].shape[0]
-        i_index = index[0]
-        j_index = index[1]
-        graph = nx.Graph()
-
-        if i_index == j_index:
-            train = window[0]
-            test = window[1]
-        else:
-            train = np.vstack([window[0], window[1]])
-            test = np.vstack([window[0], window[1]])
-
-        tree = cKDTree(train, leafsize=40)
-        edges = tree.query_ball_point(test, cutoff)
-        edge_list = [list(zip(np.repeat(idx, len(dest_list)), dest_list))
-                     for idx, dest_list in enumerate(edges)]
-
-        edge_list_flat = np.array([list(item) for sublist in edge_list for
-                                   item in sublist])
-
-        if i_index == j_index:
-            res = edge_list_flat.transpose()
-            res[0] = res[0] + i_index - 1
-            res[1] = res[1] + j_index - 1
-        else:
-            removed_elements = list()
-            for i in range(edge_list_flat.shape[0]):
-                if (edge_list_flat[i, 0] >= 0 and
-                    edge_list_flat[i, 0] <= num - 1) and \
-                    (edge_list_flat[i, 1] >= 0 and
-                     edge_list_flat[i, 1] <= num - 1) or \
-                    (edge_list_flat[i, 0] >= num and
-                     edge_list_flat[i, 0] <= 2 * num - 1) and \
-                    (edge_list_flat[i, 1] >= num and
-                     edge_list_flat[i, 1] <= 2 * num - 1) or \
-                    (edge_list_flat[i, 0] >= num and
-                     edge_list_flat[i, 0] <= 2 * num - 1) and \
-                    (edge_list_flat[i, 1] >= 0 and
-                     edge_list_flat[i, 1] <= num - 1):
-                    removed_elements.append(i)
-            res = np.delete(edge_list_flat, removed_elements,
-                            axis=0).transpose()
-            res[0] = res[0] + i_index - 1
-            res[1] = res[1] - num + j_index - 1
-        if res.shape[1] == 0:
-            res = np.zeros((2, 1), dtype=np.int)
-
-        edges = [(res[0, k], res[1, k]) for k in range(0, res.shape[1])]
-        graph.add_edges_from(edges)
-
-        # partial connected components
-
-        subgraphs = nx.connected_components(graph)
-        comp = [g for g in subgraphs]
-        return comp
+        #raise TypeError(data)
```

Comment:
I wouldn't bother with parallel LeafletFinder as it is broken at the moment #76 , #79

Reply:
Sorry I think I packed too much inside this PR. I was intended to discover the possibility to parallel LeafletFinder both among frames and inside single frame. Because for now, it only starts to work on the next frame after all the jobs are done in the current one.

So I changed this more coarsed instead of passing hundreds of jobs per frame * hundreds of frames to dask graph.

Reply:
Problem is twofold (after `AtomGroup` and everything are implemented):
- XDR and DCD format files fail to pickle the last frame:
```python
u = mda.Universe(GRO_MEMPROT, XTC_MEMPROT)
u.trajectory[4] #  last frame
pickle.loads(pickle.dumps(u.trajectory))
EOFError: Trying to seek over max number of frames
```
The major problem is trajectory._xdr.current_frame == 5 (1-based). I might need to add extra fix (and test?) to https://github.com/MDAnalysis/mdanalysis/pull/2723/files, or maybe in an individual PR? since the pickling is handled on their own
- The algorithm itself gets different results (for the same ts) with different `n_jobs` (maybe because I made some mistakes).

Reply:
The "last frame" thing is a real issue. Oops!

Don't worry about LeafletFinder at the moment, it's not really your job to fix it, and it has lots of issues. (If you need it for your own research and you have an interest in getting it working then that's a bit different but I'd still say, focus on the serialization core problem for now.)

Reply:
Just pushed a fix for the "last frame" issue.

Not `LeafletFinder` per se, but maybe a general framework to suit all conditions. e.g. 
- in each frame deserves parallelism.
- run multiple analysis on one single universe.
- run one analysis on multiple universe.
- complex dependencies between each job

A solution I can think of is to let `ParallelAnalysisBase` inherents basic methods `DaskMethodsMixin` as a [dask custom collection](https://docs.dask.org/en/latest/custom-collections.html), so that we can build complex dask graph. But I am not sure how well we can build a legit API that users do not need to care too much about the underlying implementation

Reply:
That's a good analysis of use cases and it would be useful to write this down somewhere. With PMDA so far (except LeafletFinder) we have been focusing on the simple split-apply-combine because that can be put in a simple "framework". Beyond that, it becomes difficult to do a "one size fits all" and it becomes a serious research project in CS. 

I would be happy if we had a library that allows users to easily write their own split/apply/combine type analysis and where we provide a few additional parallelized analysis that might not fit into this scheme (such as LeafletFinder).

An interesting idea that has been coming up repeatedly is to "stack" multiple analysis, i.e., run multiple `_single_frame()` sequentially so that the expensive data loading into memory time is amortized.

Finally, run one analysis on multiple universes seems to be a standard pleasingly parallel job that can make use of existing workflow management tools – I don't see what we can do directly to support it.